packages/client/hmi-client/src/utils/data-util.ts

```diff
@@ -12,81 +12,53 @@ import { Dataset, FACET_FIELDS as DATASET_FACET_FIELDS } from '@/types/Dataset';
-export const applyFacetFiltersToModels = (modelResults: Model[], filters: Filters) => {
-	if (isEmpty(filters) || isEmpty(modelResults)) {
+export const applyFacetFilters = (
+	results: Model[] | Dataset[],
+	filters: Filters,
+	resourceType: ResourceType
+) => {
+	if (isEmpty(filters) || isEmpty(results)) {
 		return;
 	}
+
 	const { clauses } = filters;
-	clauses.forEach((clause: any) => {
+	const ASSET_FACET_FIELDS: string[] =
+		resourceType === ResourceType.MODEL ? MODEL_FACET_FIELDS : DATASET_FACET_FIELDS;
+
+	clauses.forEach((clause) => {
 		const filterField: string = clause.field; // the field to filter on
 		// "filters" may include fields that belong to different types of artifacts
 		//  thus make sure to only filter models using Model fields
-		if (MODEL_FACET_FIELDS.includes(filterField)) {
+		if (ASSET_FACET_FIELDS.includes(filterField)) {
 			const filterValues = clause.values.map((v) => v.toString()); // array of values to filter upon
 			const isNot = !clause.isNot; // is the filter reversed?
-			const filteredModels = modelResults.filter(
-				(model) =>
-					// direct query against Model fields
-					filterValues.includes(model[filterField as keyof Model].toString()) === isNot
-			);
-			// use splice to filter in place
-			modelResults.splice(0, modelResults.length, ...filteredModels);
-		}
-	});
-};
 
-export const applyFacetFiltersToDatasets = (datasetResults: Dataset[], filters: Filters) => {
-	if (isEmpty(filters) || isEmpty(datasetResults)) {
-		return;
-	}
-	const { clauses } = filters;
-	clauses.forEach((clause: any) => {
-		const filterField: string = clause.field; // the field to filter on
-		// "filters" may include fields that belong to different types of artifacts
-		//  thus make sure to only filter datasets using Dataset fields
-		if (DATASET_FACET_FIELDS.includes(filterField)) {
-			const filterValues = clause.values.map((v) => v.toString()); // array of values to filter upon
-			const isNot = !clause.isNot; // is the filter reversed?
-			const filteredDatasets = datasetResults.filter(
-				(dataset) =>
-					// direct query against Dataset fields
-					filterValues.includes(dataset[filterField as keyof Dataset].toString()) === isNot
-			);
-			// use splice to filter in place
-			datasetResults.splice(0, datasetResults.length, ...filteredDatasets);
+			switch (
+				resourceType // These switch case statements avoid the typecheck issues, maybe there is a better solution?
+			) {
+				case ResourceType.MODEL:
+					results.splice(
+						0,
+						results.length,
+						...(results as Model[]).filter(
+							(asset) =>
+								filterValues.includes(asset[filterField as keyof Model].toString()) === isNot
+						)
+					);
+					break;
+				case ResourceType.DATASET:
+					results.splice(
+						0,
+						results.length,
+						...(results as Dataset[]).filter(
+							(asset) =>
+								filterValues.includes(asset[filterField as keyof Dataset].toString()) === isNot
+						)
+					);
```

Comment:
This pattern is a bit weird in the sense that a lot of other places we tend to do more of the immutable routes where functional parameters are not modified. So i.e.:

```
result = applyFacetFilters(result, ...)
```

As opposed to

```
applyFacetFilters(result, ...)
```

Here you are removing everything, then injecting items back into the same address-space...

Is there a particular reason why this is done this way? Do we have a preference more one paradigm than the other? 
@YohannParis @amostafa-uncharted

Reply:
No preference. The code was written without consistency at that time for the sake of quick implementation and perhaps other parts of the code was refactored slightly at other times resulting in such inconsistency with regards to following the im/mutable path. Happy to discuss further what you guys prefer!